Source2Gen is an executable + Linux support

Cre3per opened the pull request:
See issue #48 for details and status

Thanks for `.clang-format`, cmake, and for using assertions! It made working on this project much easier than some other projects.

es3n1n commented on include/tools/platform.h:
```diff
@@ -0,0 +1,38 @@
+
+// we're using __VA_ARGS__ to allow the use of commas, e.g. IF_WINDOWS(element,)
+#if TARGET_OS == WINDOWS
+    #define IF_WINDOWS(...) __VA_ARGS__
```
I am not a big fan of this construction, as I quickly looked at the usage of these macros. You are using them to determine offsets and such, and because of that, wouldn't it be cleaner to do something like this? https://godbolt.org/z/Tah4aKPoP

Cre3per replied:
Using constant expressions for offsets is a great idea! How about an approach using designated initialializers to make the call-sites a little more concise?

```cpp
  // using offset()
  std::uint8_t pad[offset()
                       .for_platform<e_platform::LINUX>(4)
                       .for_platform<e_platform::WINDOWS>(8)
                       .get()];
  // using designated initializers
  std::uint8_t pad2[platform_offset({.windows = 8, .linux = 4})]
```

https://godbolt.org/z/xYePfWaPT

Designated initializers might be more expensive because we're storing all possible values. Didn't benchmark.

Let me know what you think and I'll get right to it.

Also great move using an `enum class` to translate the macros, I'll be sure to use that approach in the changeset!

es3n1n replied:
Wonderful idea, yeah let's do that instead.

> Designated initializers might be more expensive because we're storing all possible values. Didn't benchmark.

I don't think it's such a big deal since most of this stuff would be done on the compile time and wouldn't even store the second (unused) value anywhere.

Cre3per replied:
Diverged slightly from my suggestion because I realized we not only have platform-specific offsets, but also vtable indexes.  
Generalized `platform_offset` to `platform_specific` and added a typecast operator because type deduction failed when using a free function. Syntax now is

```cpp
std::uint8_t pad3[platform_specific{.windows = 0x68, .linux = 0x70}];

// with .get(), when we want to use auto
constexpr auto kSchemaSystemVersion = platform_specific{.windows = 2, .linux = 1}.get();

// works with all types now
constexpr std::string_view module_file_prefix = platform_specific{.windows = "", .linux = "lib"}.get();
```

es3n1n replied:
Looks really nice, good job

es3n1n commented on src/main.cpp:
```diff
@@ -1,25 +1,19 @@
-    void OnProcessAttach(const HMODULE h_module) {
-        std::thread([h_module]() -> void { source2_gen::main(h_module); }).detach();
+int main(int argc, char* argv[]) {
+    if (source2_gen::Dump()) {
```
why not a thread?

Cre3per replied:
What would be the benefit of using a thread here?  
We no longer have `DllMain()`, but `main()`. The main thread would do nothing but spawn+join and idle while the new thread is working. Working in the main thread is simpler and we won't have to worry about what happens when an exception escapes a thread.  
I might be missing something though, curious about what you have in mind

es3n1n replied:
My bad I forgot for a second that it isn't a shared library anymore

es3n1n reviewed (changes requested):
Oh wow, good job. I took a quick look at the changes, and here are some notes. 
Not really a big fan of macros since we are coding in C++ and not C. We were trying to remove our macros a while ago, and now we are getting some new ones :)
This definitely needs lots of refactoring, but good job nonetheless

Cre3per commented on CMakeLists.txt:
```diff
@@ -12,6 +12,8 @@ if(CMAKE_SIZEOF_VOID_P EQUAL 4)
     message(FATAL_ERROR "source2gen: 32-bit targets are not supported")
 endif()
 
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
This creates `compile_commands.json` in the build directory. `compile_commands.json` is used by virtually all editors and many C++ tools on Linux to understand projects. Without it, editors will assume wrong C++ standards, fail to resolve includes, and so on. Is there an issue with this file being in the build directory?

es3n1n replied:
Good point, although aren't these IDEs setting this var by themselves? Last time I checked, I think they were setting it. Anyway it is not such a big deal, we could leave it.

Cre3per replied:
I didn't consider full IDEs, you're right, they can set this variable themselves. I use an editor for editing but build on the command line, i.e. my editor has no say in the build flags. I'm also thinking of tools like clang-tidy that run after the build and need this file.

Cre3per commented on include/tools/loader.h:
```diff
@@ -0,0 +1,70 @@
+/// Wrapping the functions in the `Platform` namespace causes compiler errors
+/// even when they're unused in the code base (Fail fast). It also provides a
+/// uniform interface for the caller.
+namespace Loader {
```
I tried using the naming convention I found in other places, I supposed I used the `Virtual` namespace as reference, which obviously can't be lower case, D'oh >.< Changed!

Cre3per commented on include/tools/platform.h:
```diff
@@ -0,0 +1,38 @@
+// Copyright (C) 2024 neverlosecc
+// See end of file for extended copyright information.
+#define WINDOWS 0
```
I got inspired by Rust's (`#[cfg(target_os = "linux")]`)[https://doc.rust-lang.org/rust-by-example/attribute/cfg.html]. I've never used this approach in C++ before and don't know how well it plays out.  
I have no preference for one or the other. I can change it to `PLATFORM_IS_*` if you like it better.

es3n1n replied:
Yeah, I am not too sure either, `#if PLATFORM == WINDOWS` looks more verbose than mine, although it's pretty much the same thing. I guess you can keep it like this, if you like this one. 

Perhaps it would make our thing 🚀 blazing fast, since It was inspired by rust 🙃

Cre3per commented:
> Oh wow, good job. I took a quick look at the changes, and here are some notes. Not really a big fan of macros since we are coding in C++ and not C. We were trying to remove our macros a while ago, and now we are getting some new ones :) This definitely needs lots of refactoring, but good job nonetheless

Thanks for the quick review and your attention to C++! I'm with you on avoiding macros as much as we can.  
We'll use a sort of `offset` class to avoid uses of `IF_LINUX/WINDOWS`. There'll be some macros for pads that exist on a single platform, unless we use `[[no_unique_address]]` and zero-sized pads, ever tried that?

I had to un-constexpr the `requiredModules` array in `startup.cpp` because I didn't want to use `LOADER_GET_MODULE_FILE_NAME()`. Do you have an idea if/how we can build module file names at compile-time without macros?

es3n1n commented:
> There'll be some macros for pads that exist on a single platform, unless we use `[[no_unique_address]]` and zero-sized pads, ever tried that? EDIT: eww, no to `[[no_unique_address]]`

Never even heard of it, lol

> I had to un-constexpr the `requiredModules` array in `startup.cpp` because I didn't want to use `LOADER_GET_MODULE_FILE_NAME()`. Do you have an idea if/how we can build module file names at compile-time without macros?

There should definitely be a way to do that. I will try to come up with an implementation, but I am a bit busy with some other projects, so this probably could take some time

cpz commented on src/startup/startup.cpp:
```diff
@@ -1,64 +1,107 @@
+            assert(handle != nullptr && "we loaded modules at startup, where did they go?");
+
+            if (const auto* pInstallSchemaBindings = loader::find_module_symbol(handle, "InstallSchemaBindings")) {
+                auto const InstallSchemaBindings = (std::uint8_t(*)(const char*, CSchemaSystem*))(pInstallSchemaBindings);
```
```            
if (auto* pInstallSchemaBindings = loader::find_module_symbol(handle, "InstallSchemaBindings")) {
                auto const InstallSchemaBindings = static_cast<std::uint8_t(*)(const char*, CSchemaSystem*)>(pInstallSchemaBindings);
```

Isnt better to use C++ cast instead of C

Cre3per replied:
Changed to `reinterpret_cast`

A `static_cast` from `void*` to function pointer is illegal. A `reinterpret_cast` and C-style cast should also be illegal afaik, but the compiler accepts it.

cpz commented on src/sdk/interfaces/tier0/IMemAlloc.cpp:
```diff
@@ -1,33 +1,45 @@
+                static const auto CMemAllocSystemInitialize = reinterpret_cast<void (*)()>(loader::find_module_symbol(tier0, "CMemAllocSystemInitialize"));
 
                 if (CMemAllocSystemInitialize) {
                     CMemAllocSystemInitialize();
```
If you want to remove while, we should try to get IMemAlloc after trying to initialize it 
`g_pMemAlloc = *reinterpret_cast<IMemAlloc**>(loader::find_module_symbol(tier0, "g_pMemAlloc"))`

cpz replied:
Also,

As I know, all these functions where the second parameter is the return value only apply to windows, so you probably need to modify these functions to work on unix
<img width="819" alt="image" src="https://github.com/neverlosecc/source2gen/assets/31280958/b9f0dcbe-3e11-4142-8d35-79436002d7c0">

Cre3per replied:
We're grabbing the game's `g_pMemAlloc` before calling `CMemAllocSystemInitialize()`, but `g_pMemAlloc` is a pointer, it will change without us having to read it again. The `g_pMemAlloc` pointer is compared against `nullptr` before returning from `MemAllocSystemInitialize()`.

cpz reviewed (commented):
otherwise looks like a good pr

cpz commented:
Also,

As I know, all these functions where the second parameter is the return value only apply to windows, so you probably need to modify these functions to work on unix
<img width="819" alt="image" src="https://github.com/neverlosecc/source2gen/assets/31280958/b9f0dcbe-3e11-4142-8d35-79436002d7c0">

cpz commented:
Providing a path for the dumper to the game is convenient if you only use the dumper on the server, but if you use it yourself, you'll get a little tired of it after a while. So on windows we'll most likely have to use Registry and from there take SteamLibrary path and hardcode the folder name for each game. Which sounds a little bit bad, but I don't know any other way. If path was provided then we'll work with it otherwise using registry + hardcoded names of folders for each game.

es3n1n commented:
> Providing a path for the dumper to the game is convenient if you only use the dumper on the server, but if you use it yourself, you'll get a little tired of it after a while. So on windows we'll most likely have to use Registry and from there take SteamLibrary path and hardcode the folder name for each game. Which sounds a little bit bad, but I don't know any other way. If path was provided then we'll work with it otherwise using registry + hardcoded names of folders for each game.

What's wrong with just placing generated SDKs in the same folder as the binary itself tho? We'll provide an optional path where to place the generated SDKs, and if it wasn't provided, we'll just drop it to the current directory

cpz commented:
> > Providing a path for the dumper to the game is convenient if you only use the dumper on the server, but if you use it yourself, you'll get a little tired of it after a while. So on windows we'll most likely have to use Registry and from there take SteamLibrary path and hardcode the folder name for each game. Which sounds a little bit bad, but I don't know any other way. If path was provided then we'll work with it otherwise using registry + hardcoded names of folders for each game.
> 
> What's wrong with just placing generated SDKs in the same folder as the binary itself tho? We'll provide an optional path where to place the generated SDKs, and if it wasn't provided, we'll just drop it to the current directory

We are talkin about game DLL's. Path is being used to get them and load them in to memory. I'm aint against dumping in current executable folder sdk, I thinkin that if I will pass path to dota 2 each time when I want to dump -- I'll break my monitor in week.

Cre3per commented:
> Also,
> 
> As I know, all these functions where the second parameter is the return value only apply to windows, so you probably need to modify these functions to work on unix <img alt="image" width="819" src="https://private-user-images.githubusercontent.com/31280958/324605630-b9f0dcbe-3e11-4142-8d35-79436002d7c0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwNjQwMDQsIm5iZiI6MTcxNDA2MzcwNCwicGF0aCI6Ii8zMTI4MDk1OC8zMjQ2MDU2MzAtYjlmMGRjYmUtM2UxMS00MTQyLThkMzUtNzk0MzYwMDJkN2MwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI1VDE2NDgyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0ZGYyNzcxM2QzOTY5MjhiYzhjMmQ4NjgzMmUxNTVkNjZhY2RkNWNjYzQ3MTFkYjI5NGNlNTJkYjdiZDA5OWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qLbmMmIiIfZtg545TgW0TSE9_DNTEBhVzcRALOLSKmg">
> 
> Edit: Just saw your comment in code <img alt="image" width="549" src="https://private-user-images.githubusercontent.com/31280958/324606603-e077690e-432f-4cd8-955d-d2920b81dabd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwNjQwMDQsIm5iZiI6MTcxNDA2MzcwNCwicGF0aCI6Ii8zMTI4MDk1OC8zMjQ2MDY2MDMtZTA3NzY5MGUtNDMyZi00Y2Q4LTk1NWQtZDI5MjBiODFkYWJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI1VDE2NDgyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0ZDU5YjU3N2MzMTNjZjVkODAxNmNhY2ZhMWZlMTQyNzhlZWZjZjgxMjI3MTIyZDZkZDJmMTA4M2U1Y2E5MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.35oNvEWdzOMfXUN9hAA4sKo0nvhDbMEcswZrh6ammko">
> 
> Isnt why we have different kSchemaSystemVersion. We need different schema system version because of support for older games than cs2 or dota2 (e.g. Dota Underlords)

Without knowing the history of source 2, I think "support older games" and "return value optimization" aren't conflicting each other. Could it be that older games or different platforms simply didn't use return value optimization? If that's the case, what we're seeing with these functions might not be caused by versioning, but changed compiler settings on Valve's side.  
If this were caused by a change in APIs, I think it'd be weird that it differs between platforms, as Valve is very likely compiling all platforms from the same source tree.  
But again, I have no experience with source 2. If you think there are different versions of this function, I'll update the comment.

Cre3per commented:
> Providing a path for the dumper to the game is convenient if you only use the dumper on the server, but if you use it yourself, you'll get a little tired of it after a while. So on windows we'll most likely have to use Registry and from there take SteamLibrary path and hardcode the folder name for each game. Which sounds a little bit bad, but I don't know any other way. If path was provided then we'll work with it otherwise using registry + hardcoded names of folders for each game.

I've had thoughts in a similar direction. `PATH` or `LD_LIBRARY_PATH` needs to contain  
`cs2/game/bin/linuxsteamrt64/` and  
`cs2/game/csgo/bin/linuxsteamrt64/`  

Source2Gen could generate the two paths itself, given a path to the game installation directory. I've added a script (`scripts/run.sh`)\* that does this on Linux. We could add a similar script for Windows.  
Linux users, in particular developers, who are the users of source2gen, will know where their game is installed and they'll know how to use their shell's history. An automatic lookup of the game directory would be of little benefit for them.

You have suggested to use the registry on Windows, should we note that idea in the linked issue so whoever restores Windows compatibility has an easier time implementing it?

\* source2gen can't set `PATH` or `LD_PRELOAD_PATH` itself because the dynamic linker reads those variables once at program launch. Changing them in the program has no effect. That's how it works on Linux at least.

es3n1n commented:
Sorry for the delays in response, I was a bit busy with some other projects, but now I am more or less free from them so you can expect faster replies from me.

> You have suggested to use the registry on Windows, should we note that idea in the linked issue so whoever restores Windows compatibility has an easier time implementing it?

Yes, please. Most likely this person would be me anyway.

> Without knowing the history of source 2, I think "support older games" and "return value optimization" aren't conflicting each other. Could it be that older games or different platforms simply didn't use return value optimization? If that's the case, what we're seeing with these functions might not be caused by versioning, but changed compiler settings on Valve's side. If this were caused by a change in APIs, I think it'd be weird that it differs between platforms, as Valve is very likely compiling all platforms from the same source tree. But again, I have no experience with source 2. If you think there are different versions of this function, I'll update the comment.

Yes, indeed. they might've changed compiler options, but since we are no valve employees we cannot know this for sure, this is why this is called a version, this variable would be used in the future if source 2 developers introduce any more changes to the compiler options/functions/anything else. This is just more convenient.

I am OK with having this comment, just add there `at least for now` and we're good :)

What are the current merge blockers for this PR? Apart from the windows support.

Cre3per commented:
> Sorry for the delays in my response(and sorry for my previous response; I was sleep-deprived, T_T). I was a bit busy with some other projects, but now I am more or less free from them, so you can expect faster replies from me.
> 
> > You have suggested to use the registry on Windows, should we note that idea in the linked issue so whoever restores Windows compatibility has an easier time implementing it?
> 
> Yes, please. Most likely this person would be me anyway.
> 
> > Without knowing the history of source 2, I think "support older games" and "return value optimization" aren't conflicting each other. Could it be that older games or different platforms simply didn't use return value optimization? If that's the case, what we're seeing with these functions might not be caused by versioning, but changed compiler settings on Valve's side. If this were caused by a change in APIs, I think it'd be weird that it differs between platforms, as Valve is very likely compiling all platforms from the same source tree. But again, I have no experience with source 2. If you think there are different versions of this function, I'll update the comment.
> 
> Yes, indeed. they might've changed compiler options, but since we are no valve employees we cannot know this for sure, this is why this is called a version, this variable would be used in the future if source 2 developers introduce any more changes to the compiler options/functions/anything else. This is just more convenient.
> 
> I am OK with having this comment; just add there `at least for now` and we're good :)

- Added a hint about the registry to the linked issue and `README.md`
- Updated the documentation of `kSchemaSystemVersion`

> What are the current merge blockers for this PR? Apart from the windows support.

- @cpz's [comments](https://github.com/neverlosecc/source2gen/pull/47#pullrequestreview-2015741031). Waiting for a re-review.
- Use of macros and missing `constexpr`. This is done from my side. We've had some good ideas in previous threads. If you've got more ideas, I'd be happy to improve the code.
- An "ok" from you guys that we can merge the Linux part untested, with a note in `README.md` stating that it is in an alpha state.
  - For CS2, source2gen generates an SDK that _looks_ right. For now, I can't run CS2 and don't have a reference SDK. I'd come back with fixes once I am a user of the SDK.
  - I haven't tested any games apart from CS2, source2gen might well crash for other games. Assuming the person who re-adds Windows support already has all the other games downloaded, I suggest they test those on Windows only. It's very possible that the `get_required_modules()` array differs between games. We could keep other Linux games "potentially broken" until CS2 is confirmed working.

It might be worth to consider adding automatic tests (Separate issue) now that source2gen is faster and doesn't need the games to be running. Even if the tests can only run on a developer machine, they'll make it safer to work on source2gen and could reduce the effort required to review pull requests.

es3n1n commented:
>     * @cpz's [comments](https://github.com/neverlosecc/source2gen/pull/47#pullrequestreview-2015741031). Waiting for a re-review.

Okie, I will let him know.

>     * Use of macros and missing `constexpr`. This is done from my side. We've had some good ideas in previous threads. If you've got more ideas, I'd be happy to improve the code.

Yeah, I will try to check it out today.

>       * For CS2, source2gen generates an SDK that _looks_ right. For now, I can't run CS2 and don't have a reference SDK. I'd come back with fixes once I am a user of the SDK.

Sure I guess, users will complain if something won't work anyway

>       * I haven't tested any games apart from CS2, source2gen might well crash for other games. Assuming the person who re-adds Windows support already has all the other games downloaded, I suggest they test those on Windows only. It's very possible that the `get_required_modules()` array differs between games. We could keep other Linux games "potentially broken" until CS2 is confirmed working.

Yeah, we will test it only on Windows. As for the required modules, we previously had an array that worked for all the source 2 games. I am too lazy to check, but if you didn't change the checked modules themselves, then we should probably be good. 

> It might be worth to consider adding automatic tests (Separate issue) now that source2gen is faster and doesn't need the games to be running. Even if the tests can only run on a developer machine, they'll make it safer to work on source2gen and could reduce the effort required to review pull requests.

Ah yes, I've been dreaming about it for a long time, adding something like GTest for tests would make the further work much easier. But that is a different story and will be implemented later.

-- Once cpz approves all the changes, I guess I will start on adding Windows support back and will most likely refactor other parts as well. This would probably need your help at some point, but we are too far from that as of now.

Cre3per commented:
> > ```
> >   * I haven't tested any games apart from CS2, source2gen might well crash for other games. Assuming the person who re-adds Windows support already has all the other games downloaded, I suggest they test those on Windows only. It's very possible that the `get_required_modules()` array differs between games. We could keep other Linux games "potentially broken" until CS2 is confirmed working.
> > ```
> 
> Yeah, we will test it only on Windows. As for the required modules, we previously had an array that worked for all the source 2 games. I am too lazy to check, but if you didn't change the checked modules themselves, then we should probably be good.

The target game doesn't need to run anymore for source2gen to work. If the game doesn't load its libraries, someone else has to. That _someone_ is source2gen.  The `get_required_modules()` array now lists all libraries that register a schema.  
For CS2, I looked at the example SDK linked on main to figure out what libraries can be dumped and listed those. If the set of libraries differs between games, we could maintain a list per-game, or load all libraries and ignore those that don't expose a `InstallSchemaBindings()` symbol.

cpz commented:
I've did some changes for compiling on Windows and dumping, there was problems with static_asserts and find_module_symbol.

find_module_symbol in loader can be changed, so it wont be based on defines, I've made like this because didnt have enough time to think, sadly atm I have limited time to do anything. 

Otherwise, seems like it works. 
We should somehow make it work based on registry on Windows cos atm for testing I was just placing needed dlls near source2gen executable.
<img width="625" alt="image" src="https://github.com/neverlosecc/source2gen/assets/31280958/32c76182-fd21-4b84-82f7-56b336b6ad94">

es3n1n reviewed (approved):
👏
